js/event.ts

```diff
@@ -0,0 +1,282 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as domTypes from "./dom_types";
+
+export const eventAttributes = new WeakMap();
+
+export class EventInit implements domTypes.EventInit {
+  bubbles = false;
+  cancelable = false;
+  composed = false;
+
+  constructor({bubbles=false, cancelable=false, composed=false} = {}) {
+    this.bubbles = bubbles;
+    this.cancelable = cancelable;
+    this.composed = composed;
+  }
+}
+
+export class Event implements domTypes.Event {
+  // Each event has the following associated flags
+  private _stopPropagationFlag = false;
+  private _stopImmediatePropagationFlag = false;
+  private _canceledFlag = false;
+  private _inPassiveListenerFlag = false;
+
+  // Property for objects on which listeners will be invoked
+  private _path: domTypes.EventPath[] = [];
+
+  constructor(type: string, eventInitDict?: domTypes.EventInit) {
+    eventAttributes.set(this, {
+      type,
+      bubbles: eventInitDict && eventInitDict.bubbles || false,
+      cancelable: eventInitDict && eventInitDict.cancelable || false,
+      composed: eventInitDict && eventInitDict.composed || false,
+      currentTarget: null,
+      eventPhase: domTypes.EventPhase.NONE,
+      isTrusted: false,
+      target: null,
+      timeStamp: Date.now(),
+    });
+  }
+
+  get bubbles(): boolean {
+    if (eventAttributes.has(this)) {
```

Comment:
Do we really need this check?  There is no logical way that you would not have an entry in the map.  I am all for safety, but this seems obsessive and even if it did, the thrown error doesn't really add anything.

Reply:
I worked with @bathos on this implementation so I would say the check is worthwhile but _need_ is a strong word. Flipping it around - does it need to be removed?

Reply:
It’s how branded objects, including all(?) Web IDL objects and many ES intrinsics, work. Their methods throw a TypeError when invoked with incompatible receivers (i.e., any object which was not initialized by the constructor). For example the expression `Event.prototype.bubbles` throws.

I think whether it’s necessary is another question about overall intention. If the goal is to implement the spec, it would seem to be necessary\*. If it’s to implement a subset of the spec, it may not be. If it’s the latter, I think it would help if the parameters were codified somewhere (e.g. ”don’t implement any branding. model private state using the following pattern instead: ...” etc) so people can stay on the same page. If it’s the former though, dotting all the i’s and crossing all the t’s one at a time is probably too much boilerplate and introduces too many places for something to go wrong, so abstracting it away into a set of common Web IDL tools would help.

> \* This throwing behavior is deliberately employed in real life stuff btw. Cross realm brand checking is only possible by leveraging side-effect-free branded methods of intrinsic classes. For example:
>
> ```js
> const test = Object.getOwnPropertyDescriptor(Event.prototype, 'bubbles').get;
> const { apply } = Reflect;
>
> export function isEvent(value) {
>   try {
>     apply(bubbles, value, []);
>     return true;
>   } catch {
>     return false;
>   }
> }
> ```
>
> Such a function will tell you if something is a validly initialized Event object regardless of what realm it originated in.

Reply:
@acconrad @bathos-wistia the goals are not to be obsessively pedantic about compliance to the spec, especially with an assumption that consuming code is going to be type checked with TypeScript.  It was unawareness of alignment to the spec that made me think that the whole logic was redundant.

> dotting all the i’s and crossing all the t’s one at a time is probably too much boilerplate and introduces too many places for something to go wrong, so abstracting it away into a set of common Web IDL tools would help

That is a good point, if we wanted to preserve this functionality, we should have a helper function that looks it up.  Something like:

```ts
export function getPrivateValue<
  K extends object,
  V extends object,
  W extends keyof V
>(instance: K, weakMap: WeakMap<K, V>, key: W): V[W] {
  if (weakMap.has(instance)) {
    return weakMap.get(instance)[key];
  }
  throw new TypeError("Illegal invocation");
}
```

And then the property getters can just be:

```ts
get bubbles(): boolean {
  return getPrivateValue(this, eventAttributes, "bubbles");
}
```

We just don't have a good home for it at the moment, it could just get added to `util.ts` I guess.  If we can do something like that, then I would be more comfortable and we can just test that behaviour of throwing in one spot instead of having to try to test that logic on every "private getter".

Reply:
Yep, that’s the pattern usually used when implementing host APIs in ES.

The TypeScript factor makes sense. It seems totally reasonable to use the spec as only a high-level guideline here. Real world intended use cases and practical considerations ought to dictate your approach. However I’m gonna take the bait the first sentence: Something may be spec compliant or not spec compliant. There isn’t a third tier, “pedantic compliance” :) When people do aim for spec compliance, it isn’t usually for kicks. There are (typically) numerous reasons / benefits.

Even if compliance were a goal here, it wouldn’t make much sense for hitting on the first pass in a single component like this. Many parts of the web platform are interconnected and share algorithms, so you’d probably want to do a fair amount of high level planning / start from more basic building blocks if spec compliance were the goal. When Adam asked me about spec conformance, I didn’t have the whole context. I think I have a better sense of what you’re looking for here now.

Reply:
I really like this implementation! Added